src/awkward/_connect/numpy.py

```diff
@@ -35,15 +40,47 @@ def convert_to_array(layout, args, kwargs):
 implemented = {}
 
 
-def _to_rectilinear(arg):
-    backend = ak._backends.backend_of(arg, default=None)
-    # We have some array-like object that our backend mechanism understands
-    if backend is not None:
-        return ak.operations.to_numpy(arg)
+def _find_backends(args: Iterable) -> Backend:
+    """
+    Args:
+        args: iterable of objects to visit
+
+    Returns the backend of the first layout or high-level Awkward object encountered.
+    """
+    stack = collections.deque(args)
+    while stack:
+        arg = stack.popleft()
+        # If the argument declares a backend, easy!
+        backend = backend_of(arg, default=None)
+        if backend is not None:
+            yield backend
+        # Is this object something we already associate with an nplike?
+        elif isinstance(arg, (tuple, list)):
+            stack.extend(arg)
```

Comment:
I don't love that we have to do two passes over the arguments list here. There are other ways to do this, but this seems like the most elegant approach. My feeling is that we want to provide native implementations for these functions if they're commonly needed, and further that we are mainly concerned with array-bound performance, so the fact that we visit the args twice is not the end of the world.

Reply:
I think that doing two passes over all of the arguments is fine. In fact, I probably wouldn't have bothered with `collections.deque` to make a stack, since the built-in list is already stack-like, with `append` and `pop`. Oh, I see: `popleft`. Well, built-in list is still good for that.

(The number of arguments in a function call is never very large.)